File: dsa_leetcode_beginner/0057-insert-interval/0057-insert-interval.js

function insert(intervals, newInterval) {
    let left = [], right = [];
    let [start, end] = newInterval;

    for (let i of intervals) {
        if (i[1] < start) {
            left.push(i);
        } else if (i[0] > end) {
            right.push(i);
        } else {
            start = Math.min(start, i[0]);
            end = Math.max(end, i[1]);
        }
    }

    return [...left, [start, end], ...right];
}

// Example usage:
let intervals = [[1, 3], [6, 9]];
let newInterval = [2, 5];
console.log(insert(intervals, newInterval));

File: dsa_leetcode_beginner/0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals, newInterval) -> List[List[int]]:
        left, right = [], []
        start, end = newInterval

        for i in intervals:
            # i ends before new_interval starts
            if i[1] < start:  
                left.append(i)

            # i starts after new_interval ends
            elif i[0] > end:  
                right.append(i)

            # i overlaps with new_interval
            else:  
                start = min(start, i[0])
                end = max(end, i[1])

        return left + [[start, end]] + right


#Question: https://leetcode.com/problems/insert-interval
#Blog: https://blog.unwiredlearning.com/insert-interval